Soportar DVFS y subsistema térmico en la GPU

Este parche, de creación propia, soporta la refrigeración pasiva de la GPU mediante el cambio de su frecuencia de operación. Naturalmente, también hace posible
cambiar la frecuencia de la GPU independientemente de la condición térmica de la Pine.

Las frecuencias y voltajes han sido extraídos del BSP original de Allwinner: https://github.com/Allwinner-Homlet/H6-BSP4.9-linux/blob/master/arch/arm64/boot/dts/sunxi/sun50iw6p1.dtsi

Signed-off-by: AlexTMjugador <AlexTMjugador@users.noreply.github.com>
---
diff -pur linux.orig/arch/arm64/boot/dts/allwinner/sun50i-h6.dtsi linux/arch/arm64/boot/dts/allwinner/sun50i-h6.dtsi
--- linux.orig/arch/arm64/boot/dts/allwinner/sun50i-h6.dtsi	2019-08-19 20:56:12.778926525 +0200
+++ linux/arch/arm64/boot/dts/allwinner/sun50i-h6.dtsi	2019-08-19 21:25:01.882871369 +0200
@@ -225,6 +225,87 @@
 			clock-names = "core", "bus";
 			resets = <&ccu RST_BUS_GPU>;
 			status = "disabled";
+			operating-points-v2 = <&gpu_opp_table>;
+			#cooling-cells = <2>;
+		};
+
+		gpu_opp_table: opp_table {
+			compatible = "operating-points-v2";
+
+			opp@216000000 {
+				opp-hz = /bits/ 64 <216000000>;
+				opp-microvolt = <810000>;
+			};
+
+			opp@264000000 {
+				opp-hz = /bits/ 64 <264000000>;
+				opp-microvolt = <810000>;
+			};
+
+			opp@312000000 {
+				opp-hz = /bits/ 64 <312000000>;
+				opp-microvolt = <810000>;
+			};
+
+			opp@336000000 {
+				opp-hz = /bits/ 64 <336000000>;
+				opp-microvolt = <810000>;
+			};
+
+			opp@360000000 {
+				opp-hz = /bits/ 64 <360000000>;
+				opp-microvolt = <820000>;
+			};
+
+			opp@384000000 {
+				opp-hz = /bits/ 64 <384000000>;
+				opp-microvolt = <830000>;
+			};
+
+			opp@408000000 {
+				opp-hz = /bits/ 64 <408000000>;
+				opp-microvolt = <840000>;
+			};
+
+			opp@420000000 {
+				opp-hz = /bits/ 64 <420000000>;
+				opp-microvolt = <840000>;
+			};
+
+			opp@432000000 {
+				opp-hz = /bits/ 64 <432000000>;
+				opp-microvolt = <860000>;
+			};
+
+			opp@456000000 {
+				opp-hz = /bits/ 64 <456000000>;
+				opp-microvolt = <870000>;
+			};
+
+			opp@504000000 {
+				opp-hz = /bits/ 64 <504000000>;
+				opp-microvolt = <890000>;
+			};
+
+			opp@540000000 {
+				opp-hz = /bits/ 64 <540000000>;
+				opp-microvolt = <910000>;
+			};
+
+			opp@576000000 {
+				opp-hz = /bits/ 64 <576000000>;
+				opp-microvolt = <930000>;
+			};
+
+			opp@624000000 {
+				opp-hz = /bits/ 64 <624000000>;
+				opp-microvolt = <950000>;
+			};
+
+			opp@756000000 {
+				opp-hz = /bits/ 64 <756000000>;
+				opp-microvolt = <1040000>;
+			};
 		};
 
 		syscon: syscon@3000000 {
@@ -869,6 +950,41 @@
 			polling-delay-passive = <250>;
 			polling-delay = <1000>;
 			thermal-sensors = <&ths 1>;
+
+			cooling-maps {
+				map0 {
+					trip = <&gpu_alert0>;
+					cooling-device = <&gpu THERMAL_NO_LIMIT THERMAL_NO_LIMIT>;
+				};
+
+				map1 {
+					trip = <&gpu_alert1>;
+					cooling-device = <&gpu THERMAL_NO_LIMIT THERMAL_NO_LIMIT>;
+				};
+			};
+
+			trips {
+				gpu_alert0: gpu_alert0 {
+					/* milliCelsius */
+					temperature = <75000>;
+					hysteresis = <2000>;
+					type = "passive";
+				};
+
+				gpu_alert1: gpu_alert1 {
+					/* milliCelsius */
+					temperature = <90000>;
+					hysteresis = <2000>;
+					type = "hot";
+				};
+
+				gpu_crit: gpu_crit {
+					/* milliCelsius */
+					temperature = <1000000>;
+					hysteresis = <2000>;
+					type = "critical";
+				};
+			};
 		};
 	};
 };
